### PR TITLE
Demo product refactor to typescript

### PR DESCRIPTION
**Pull request by developerium**
Hi Guys
I want to contribute to the refactoring demo app to typescript.
I want to refactor `examples/demo/src/products` folder. Please tell me if someone else is already working on it.

Stay safe

### CODE REVIEW

File: examples/demo/src/products/GridList.tsx [hunk 1]
```diff
@@ -1,12 +1,34 @@
...
 import GridListTileBar from '@material-ui/core/GridListTileBar';
 import { makeStyles } from '@material-ui/core/styles';
 import withWidth from '@material-ui/core/withWidth';
 import { Link } from 'react-router-dom';
 import { NumberField } from 'react-admin';
-import { linkToRecord } from 'ra-core';
+import { linkToRecord, Identifier } from 'ra-core';
+import { Product } from '../types';
+
+enum Width {
```
Comment (fzaninotto) [inline examples/demo/src/products/GridList.tsx]: No, use the type exported by Material-ui for this one

File: examples/demo/src/products/GridList.tsx [hunk 2]
```diff
@@ -33,18 +55,18 @@ const useStyles = makeStyles(theme => ({
...
+const getColsForWidth = (width: Width) => {
     if (width === 'xs') return 2;
     if (width === 'sm') return 3;
     if (width === 'md') return 4;
     if (width === 'lg') return 5;
     return 6;
 };
 
-const times = (nbChildren, fn) =>
+const times = (nbChildren: number, fn: TimesCallback) =>
```
Comment (fzaninotto) [inline examples/demo/src/products/GridList.tsx]: We really don't need a named type for the second argument.

File: examples/demo/src/products/GridList.tsx [hunk 3]
```diff
@@ -1,12 +1,34 @@
...
+interface GridListProps {
+    ids: Identifier[];
+    data: { [key: string]: Product };
+    basePath: string;
+    width: Width;
+    nbItems: number;
+}
+
+interface Props extends GridListProps {
+    loaded: boolean;
```
Comment (fzaninotto) [inline examples/demo/src/products/GridList.tsx]: No, there are way more props injected to GridList

File: examples/demo/src/products/GridList.tsx [hunk 4]
```diff
@@ -106,7 +128,7 @@ const LoadedGridList = ({ ids, data, basePath, width }) => {
     );
 };
 
-const GridList = ({ loaded, ...props }) =>
+const GridList: FC<Props> = ({ loaded, ...props }) =>
     loaded ? <LoadedGridList {...props} /> : <LoadingGridList {...props} />;
 
-export default withWidth()(GridList);
+export default withWidth()(GridList as any);
```
Comment (fzaninotto) [inline examples/demo/src/products/GridList.tsx]: Not good. If you have to use any, try to look for an alternative solution. For instance, not using `withWidth`, but the new hooks provided by material-ui.

File: examples/demo/src/products/GridList.tsx (hunk 1 above)
Comment (developerium) [reply examples/demo/src/products/GridList.tsx:12]: Done

File: examples/demo/src/products/GridList.tsx (hunk 2 above)
Comment (developerium) [reply examples/demo/src/products/GridList.tsx]: Done

File: examples/demo/src/products/GridList.tsx (hunk 4 above)
Comment (developerium) [reply examples/demo/src/products/GridList.tsx]: Done

**Comment by developerium** [comment]
I need a little help, I'm facing an issue like this:
https://github.com/mui-org/material-ui/issues/9106

On **demo/src/products/GridList.tsx:92**, `Link` is passes in `component` prop of **GridListTile**, and also `to` is used with string, and we get the error `TS2769: No overload matches this call`.

I tried the below solutions but didn't work:
https://github.com/mui-org/material-ui/issues/9106#issuecomment-406598976
https://github.com/mui-org/material-ui/issues/9106#issuecomment-343732618

Can you direct me in the right path?
@fzaninotto @djhi

**Comment by josephktcheung** [comment]
> I need a little help, I'm facing an issue like this:
> [mui-org/material-ui#9106](https://github.com/mui-org/material-ui/issues/9106)
> 
> On **demo/src/products/GridList.tsx:92**, `Link` is passes in `component` prop of **GridListTile**, and also `to` is used with string, and we get the error `TS2769: No overload matches this call`.
> 
> I tried the below solutions but didn't work:
> [mui-org/material-ui#9106 (comment)](https://github.com/mui-org/material-ui/issues/9106#issuecomment-406598976)
> [mui-org/material-ui#9106 (comment)](https://github.com/mui-org/material-ui/issues/9106#issuecomment-343732618)
> 
> Can you direct me in the right path?
> @fzaninotto @djhi

I believe it was a bug of material-ui which should be fixed in 4.9.13 https://github.com/mui-org/material-ui/commit/701e3ad76b788a50ea83aeb8516ed303c2435bd0, which allows `GridListTile` to accept generic component as `component` prop. I tried upgrading `@material-ui/core` to 4.9.13 but failed to compile `ra-ui-materialui`.

@fzaninotto would react-admin consider upgrading `@material-ui/core` in the near future?

**Comment by developerium** [comment]
Thank you @josephktcheung for bringing this up.
I've bumped the `@material-ui/core` in `examples/demo/package.json` to latest and I can confirm that this problem was fixed, but as you mentioned, there were lot's of other new issues related to the theme and I thought that this was out of the scope of this pull request, so I want to just ignore type check on that specific line and continue with the refactoring.

File: examples/demo/src/products/GridList.tsx (hunk 3 above)
Comment (developerium) [reply examples/demo/src/products/GridList.tsx]: Done

**Comment by fzaninotto** [comment]
Could you please rebase on `next`? that's where we'll add the types to the demo app.

**Comment by MohammedFaragallah** [comment]
Hello, @developerium are you still working on this?

**Comment by developerium** [comment]
Hi @MohammedFaragallah 
I'm actually swamped with work and I think I'll be busy for a while, so I don't mind if anybody wants to work on this.

**Comment by fzaninotto** [comment]
The demo is now entirely migrated to TypeScript.

**Pull request closed without merging**